src/treelearner/serial_tree_learner.cpp

```diff
@@ -687,4 +697,46 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
   }
 }
 
+void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
+  FeatureHistogram* histogram_array_;
+  if (!histogram_pool_.Get(leaf, &histogram_array_)) {
+    Log::Warning(
+        "Get historical Histogram for leaf %d failed, will skip the "
+        "``RecomputeBestSplitForLeaf``",
+        leaf);
+    return;
+  }
+  double sum_gradients = split->left_sum_gradient + split->right_sum_gradient;
+  double sum_hessians = split->left_sum_hessian + split->right_sum_hessian;
+  int num_data = split->left_count + split->right_count;
+
+  std::vector<SplitInfo> bests(share_state_->num_threads);
+  LeafSplits leaf_splits(num_data);
+  leaf_splits.Init(leaf, sum_gradients, sum_hessians);
+
+  OMP_INIT_EX();
+// find splits
+#pragma omp parallel for schedule(static)
+  for (int feature_index = 0; feature_index < num_features_; ++feature_index) {
+    OMP_LOOP_EX_BEGIN();
+    if (!col_sampler_.is_feature_used_bytree()[feature_index]) {
+      continue;
+    }
+    if (!histogram_array_[feature_index].is_splittable()) {
+      continue;
+    }
+    const int tid = omp_get_thread_num();
+    int real_fidx = train_data_->RealFeatureIndex(feature_index);
+    ComputeBestSplitForFeature(
+        histogram_array_, feature_index, real_fidx,
+        true,
```

Comment:
@guolinke should I be passing elements of col_sampler_.GetByNode() (that I would call at the start of the function) here instead of true now? That way I could remove the warning about monotone constraints and feature fractions not working together? Thanks

Reply:
Not, the `col_sampler_.GetByNode()` returns the different results in every call.